### PR TITLE
(maint) Fixes for custom bill-of-materials paths

### PR DESCRIPTION
**Pull request by underscorgan**
When using a custom project bill-of-materials path for precompiled archives, the bill-of-materials file was not correctly removed from the compiled archive, and thus conflicts with projects using the archive with the same custom bill-of-materials path.

The files added to the tarball should include the path to the bill-of-materials, not just the default bill-of-materials location.

### CODE REVIEW

File: lib/vanagon/platform.rb [hunk 1]
```diff
@@ -469,13 +469,27 @@ def generate_compiled_archive(project)
...
+      # BOM is moved to the custom path. So, when cleaning up BOMs for non-custom
+      # paths we just want to remove the BOM at the top level of the tarball.
+      # But, if we have a custom BOM path we want to move it back to where it
+      # was prior to the Makefile execution so we can preserve it as an artifact
+      # but not leave it to conflict if it's installed in the same custom path
+      # as a project using this archive.
+      bill_of_materials_command = 'rm -f bill-of-materials'
+      if project.bill_of_materials
+        bill_of_materials_command = "mv .#{project.bill_of_materials.path}/bill-of-materials ../.."
+      end
```
Comment (mckern) [inline lib/vanagon/platform.rb]: twocents_1914


Consider this as a more idiomatic Ruby variable assignment:

```ruby
bill_of_materials_command = if project.bill_of_materials
                              "mv .#{project.bill_of_materials.path}/bill-of-materials ../.."
                            else
                              'rm -f bill-of-materials'
                            end
```

(though I think Rubocop might want it aligned under `bill_of_materials_command`, not under `if`)

File: lib/vanagon/project.rb [hunk 2]
```diff
@@ -525,7 +525,13 @@ def has_services?
     #
     # @return [Array] all the files and directories that should be included in the tarball
     def get_tarball_files
-      files = ['file-list', 'bill-of-materials']
+      files = ['bill-of-materials']
+
+      if bill_of_materials
+        files = ["#{bill_of_materials.path}/bill-of-materials"]
+      end
```
Comment (mckern) [inline lib/vanagon/project.rb]: Samesies ^^

File: lib/vanagon/platform.rb (hunk 1 above)
Comment (mwaggett) [reply lib/vanagon/platform.rb]: `bill_of_materials_command = project.bill_of_materials ? "mv .#{project.bill_of_materials.path}/bill-of-materials ../.." : 'rm -f bill-of-materials'` seems more idiomatic to me?
but also, I'm totally fine with how it is rn.

**Comment by mwaggett** [comment]
I'm honestly a little confused about what exactly this is fixing, but I trust you so

**Comment by underscorgan** [comment]
@mwaggett there were a few places in the code where we weren't paying attention to the custom directory the bill-of-materials file is stored under when we were operating on the bill-of-materials. This updates those places to use the directory if it's set.

This issue cropped up when both a runtime and parent project have the same custom path. Without these fixes, in that case the runtime bill-of-materials would attempt to overwrite the parent project bill-of-materials and cause build failures. This makes custom bills of materials handled consistently with the standard bill of materials.

**Comment by underscorgan** [comment]
And for the assignment style comments. I can update to use @mckern 's suggestion if there are strong feelings, but I don't have any for that 🤷‍♀️ . I'm generally 👎 on using selectors, they tend to make code more confusing / harder to read so I don't want to use that @mwaggett.

**Comment by underscorgan** [comment]
@mwaggett The method where it was being removed and is now being moved out of the way is a compiled archive (like runtime) specific method. For compiled archives, they are formatted such that they're intended to be unpacked at the root level of the filesystem. For those jobs, the bill of materials is shipped alongside the archive rather than inside of it so it doesn't litter the filesystem.

We're moving it so it ends up in the right place to get shipped alongside the archive. This relates to https://github.com/puppetlabs/vanagon/blob/master/resources/Makefile.erb#L42-L44. With custom bill of materials paths the file is moved from its default location during the makefile execution, which means there is no bill of materials at the top level build directory if there's a custom path.


tl;dr - for compiled archive generation we were incorrectly shipping a bill-of-materials in the archive where we shouldn't have been. Moving it removes it from the archive while maintaining it as an artifact to ship alongside the compiled archive.

**Comment by mwaggett** [comment]
ahhhh gotcha, thank you for explaining!
